### DISCUSSION TITLE
Support Incremental Full Route Cache on Dynamic Segments

### DISCUSSION THREAD

**Discussion by awinogrodzki** [vercel/next.js#57961] [App Router]
Hello,

Before migrating to App Router, I was using [Incremental Static Regeneration](https://nextjs.org/docs/pages/building-your-application/data-fetching/incremental-static-regeneration) to generate static pages on the fly.

I am running a CMS app with routes similar to `/blog/[slug]`, which means that I cannot generate pages during build time and cannot use `Full Route Cache` that comes with Next.js 13/14

I have a single API call during blog post render cycle, which is successfully cached.

Even with this API call coming from cache, my throughput dropped from `250-300 requests/s` to at most `10 requests/s` on production instances

This is a drastic change that has been reflected in infrastructure costs. 

It's basically impossible to revert to [Incremental Static Regeneration](https://nextjs.org/docs/pages/building-your-application/data-fetching/incremental-static-regeneration) at this moment as most of the new features use App Router.

Are there any plans to bring Full Route Cache to Dynamic Segments?

**Comment by HarleySalas** [comment]
you can *optionally use `generateStaticParams` at build time, to ensure that existing pages are already ready and waiting for users after each build. After that, any updates can be revalidated and additions will be cached at runtime as well.

You just make a page `async` as a server component and after it's fetched for the first time, it will be statically cached by default.

 Is this behavior missing anything, or am I missing something?

**Reply by awinogrodzki** [reply]
Thanks for the prompt answer @HarleySalas!

In my app, users can add new posts dynamically.

Let's assume I use this `generateStaticParams` function from [official example](https://nextjs.org/docs/app/api-reference/functions/generate-static-params) in `/blog/[slug]/page.tsx`:

```tsx
export async function generateStaticParams() {
  const posts = await fetch('https://api.blog-example.com/posts').then((res) => res.json()) 
 
  return posts.map((post) => ({
    slug: post.slug,
  }))
}
```
Example scenario:
`https://api.blog-example.com/posts` might return two slugs at build time: `slug-1` and `slug-2`.

After build, user adds another page to the blog: `slug-3`.

Will revalidating `/blog/slug-3` statically cache the new post, or would app have to be rebuilt in order to use static cache for `/blog/slug-3`?

**Reply by awinogrodzki** [reply]
Just to give you an overview of the scale I am dealing with:

My app can support N number of blogs, and each blog can have M number of posts managed by multiple users.

I don't really look to statically cache anything during build time, because it can drastically increase resulting docker image.

What I am looking for is incremental static regeneration, so basically, if the behaviour is as you described, I can do the following inside `/blog/[slug]/page.tsx`:

```tsx
export async function generateStaticParams() { 
  return [];
}
```

Will this enable incremental static regeneration for `/blog/[slug]/page.tsx` or am I missing something? 

I think I am missing something, because even with 
```tsx
export async function generateStaticParams() { 
  return [];
}
```
the page seems to be rendered on every request

**Reply by HarleySalas** [reply]
If you aren't generating static params at build time, you don't even need to have the function at all. By default, all pages are static.

`generateStaticParams` simply requests and fetches data ahead of time, so that the first user visiting the page after a build will not have to wait.

I'm not sure what *exactly* you are doing in your `[slug]` route, so I cannot confirm if it is still static, or if you've done something that has caused it to be dynamic.

If you're doing this testing in dev mode, it will certainly appear as if it's being rendered on every request, as it does render it on every request in dev mode, regardless.

If you want to test the behavior of your application, you should build and run it on your device. When the build completes, it will show you a list of the pages that are static and which are dynamic.

You can also experiment with the revalidation as well. Try adding an article, you'll see that it shows up right away. Make an update to the article, but don't revalidate it and it's changes *should* not be reflected yet, until the next request after your `ttl` is reached, or when you revalidate it.

Once you go through that, if you haven't already, if it's still not working as you expect, please share the relevant code from your `page.tsx` file inside of your `[slug]` folder.

**Reply by awinogrodzki** [reply]
I created this repository to illustrate my problem and deployed it to Vercel: https://github.com/awinogrodzki/static-pages-test

https://static-pages-test.vercel.app/static-page

The app features two routes `/static-page` and `/dynamic-page/[segment]`

#### Static page

Static page is defined as follows: https://github.com/awinogrodzki/static-pages-test/blob/main/app/static-page/page.tsx

The preview can be seen here: https://static-pages-test.vercel.app/static-page

The timestamp is not changing upon refresh, which means page is not re-rendered. All working as expected.

#### Dynamic page

The code for dynamic page: https://github.com/awinogrodzki/static-pages-test/blob/main/app/dynamic-page/%5Bsegment%5D/page.tsx

The preview can be seen here: https://static-pages-test.vercel.app/dynamic-page/foo-bar

Because of dynamic segment, page is marked as `dynamic` in build output log. When you refresh the preview, the timestamp is changing on each request, which means page is re-rendered on each and every request, despite being called with the same `foo-bar` segment.


I would expect `/dynamic-page/foo-bar` to be incrementally cached, as it would if we define `generateStaticParams`:

```
export async function generateStaticParams() {
  return [];
}
```

What is interesting – when you add this export
```tsx
export async function generateStaticParams() {
  return [];
}
```
to dynamic page: https://github.com/awinogrodzki/static-pages-test/blob/main/app/dynamic-page/%5Bsegment%5D/page.tsx

build output log now marks `/dynamic-page/[segment]` as `SSG` and everything starts working as expected.


It seems that I was missing this export.

Also, it proves that if you don't define dummy `generateStaticParams`, page will not be cached.

**Reply by phips28** [reply]
Thx for that hint! We also have those dynamically added blog posts (without redeploy) - and were looking for a way to cache those pages after first user hit the page.
```javascript
export async function generateStaticParams() {
  return [];
}
```

That should be documented somewhere, or did I miss it in the docs? 😄  (they are huge)

**Reply by jhp0115** [reply]
When I wrote `export const dynamicParams = true;` in Page.tsx file, I found dynamic routes can also be cached. (tested with next.js 14.1.4)

```tsx
export const dynamicParams = true;


export async function generateStaticParams() { /* ... */ }
export default async function Page(/* ... */) { /* ... */ }
```

https://nextjs.org/docs/app/building-your-application/upgrading/app-router-migration#replacing-fallback

In this link, I found the following sentence:

"With `dynamicParams` set to `true` (the default), when a route segment is requested that hasn't been generated, it will be server-rendered and **cached**."

When I omit this in Page.tsx, dynamic routes were not cached, but when I add this and explicitly set to `true`, dynamic routes were cached.

**Reply by phips28** [reply]
Good to know @jhp0115 . does that also work when you remove `generateStaticParams()` ?

**Reply by jhp0115** [reply]
When I remove `generateStaticParams()`, it didn't work.

But I think returning an empty array in `generateStaticParams()` can be a way when no params are determined.

**Reply by Sammeeey** [reply]
If this works, that's great!
How can I test whether my pages get cached on the server?

Looks like this for me now
```
Route (app)                              Size     First Load JS
┌ ○ /                                    2.04 kB        92.7 kB
├ ○ /_not-found                          870 B          84.8 kB
├ ● /[hpId]        <<<----😍
...


ƒ Middleware                             763 kB

○  (Static)   prerendered as static content
●  (SSG)      prerendered as static HTML (uses getStaticProps)
λ  (Dynamic)  server-rendered on demand using Node.js
```
... using
```
export const generateStaticParams = async () => {
  return [];
};
```

**Reply by jhp0115** [reply]
You can find the static html file in `.next/server/app/...` after you access a new route..!

For an example, if your target route is `/hpid-1234`, `hpid-1234.html`, `hpid-1234.meta` and `hpid-1234.rsc` will be generated in `.next/server/app/` after you run the server and access the route.

(I checked returning an empty array in generateStaticParams() works.)

**Reply by magnusriga** [reply]
@jhp0115 Have you figured out how Streaming and Suspense plays into this? Assuming you do not have PPR turned on, will the cached page (the one cached on first visit) be served and suspense play no role? Will anything re-render?

Without PPR, are layout and pages still cached separately, so you can call a Dynamic API like `headers()` in the page, and still get the cached layout served right away?

**EDIT**: On further inspection it seems the route is not cached at all, if a Dynamic API appears anywhere (layout or page). Which means fetches from remote cache will still look visibly slow without PPR, but the if they are wrapped in suspense then at least the rest will be fast.
